protocol/x/vault/keeper/withdraw.go

```diff
@@ -0,0 +1,119 @@
+package keeper
+
+import (
+	"math/big"
+
+	errorsmod "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/dydxprotocol/v4-chain/protocol/lib"
+	"github.com/dydxprotocol/v4-chain/protocol/lib/vault"
+	"github.com/dydxprotocol/v4-chain/protocol/x/vault/types"
+)
+
+// GetVaultWithdrawalSlippagePpm returns the slippage that should be incurred from the specified
+// vault on withdrawing `sharesToWithdraw` shares.
+// For example, if `sharesToWithdraw = 100` and `0.2` is returned, it means that withdrawing
+// 100 shares has a 20% slippage for the given `vaultId`.
+//
+// Slippage is calculated as `min(simple_slippage, estimated_slippage)` where:
+// - simple_slippage = leverage * initial_margin
+// - estimated_slippage = spread * (1 + average_skew) * leverage
+//   - average_skew = integral / (posterior_leverage - leverage)
+//   - integral = skew_antiderivative(skew_factor, posterior_leverage) -
+//     skew_antiderivative(skew_factor, leverage)
+//   - posterior_leverage = leverage / (1 - withdrawal_portion)
+//     = leverage / (1 - shares_to_withdraw / total_shares)
+//     = leverage * total_shares / (total_shares - shares_to_withdraw)
+//
+// To simplify above formula, let l = leverage, n = total_shares, m = shares_to_withdraw
+//
+//	estimated_slippage
+//	= spread * (1 + integral / (posterior_leverage - leverage)) * leverage
+//	= spread * (1 + integral * (n - m) / (ln - l(n - m))) * l
+//	= spread * (1 + integral * (n - m) / lm) * l
+//	= spread * (l + integral * (n - m) / m)
+func (k Keeper) GetVaultWithdrawalSlippage(
```

Comment:
nit: Might as well convert back to a big.Int to return ppm slippage here rather than outside of this function?

Reply:
i was thinking that returning a rational number can allow parent functions to yield more accurate results (e.g. `equity * (1 - slippage)` can be more accurate if slippage isn't yet rounded to a big.Int yet)